gssapi/mechs.py

```diff
@@ -0,0 +1,205 @@
+import six
+
+from gssapi.raw import oids as roids
+from gssapi._utils import import_gssapi_extension
+from gssapi.raw import misc as rmisc
+from gssapi import _utils
+
+rfc5587 = import_gssapi_extension('rfc5587')
+rfc5801 = import_gssapi_extension('rfc5801')
+
+
+class Mechanism(roids.OID):
+    """
+    A GSSAPI Mechanism
+
+    This class represents a mechanism and centralizes functions dealing with
+    mechanisms and can be used with any calls.
+
+    It inherits from the low-level GSSAPI :class:`~gssapi.raw.oids.OID` class,
+    and thus can be used with both low-level and high-level API calls.
+    """
+    def __new__(cls, cpy=None, elements=None):
+        return super(Mechanism, cls).__new__(cls, cpy, elements)
+
+    @property
+    def names(self):
+        """
+        Get the set of name types supported by this mechanism.
+        """
+        return rmisc.inquire_names_for_mech(self)
+
+    @property
+    def _saslname(self):
+        if rfc5801 is None:
+            raise NotImplementedError("Your GSSAPI implementation does not "
+                                      "have support for RFC 5801")
+        return rfc5801.inquire_saslname_for_mech(self)
+
+    @property
+    def _attrs(self):
+        if rfc5587 is None:
+            raise NotImplementedError("Your GSSAPI implementation does not "
+                                      "have support for RFC 5587")
+
+        return rfc5587.inquire_attrs_for_mech(self)
+
+    def __str__(self):
+        if issubclass(str, six.text_type):
+            # Python 3 -- we should return unicode
+            return self.__bytes__().decode(_utils._get_encoding())
+        else:
+            return self.__bytes__()
+
+    def __unicode__(self):
+        return self.__bytes__().decode(_utils._get_encoding())
+
+    def __bytes__(self):
+        base = self.dotted_form
+        if rfc5801 is not None:
+            base = self._saslname.mech_name
+
+        if issubclass(str, six.text_type):
+            base = bytes(base, _utils._get_encoding())
+        else:
+            base = bytes(base)
```

Comment:
Since `dotted_form` returns a string, this needs to be cast to `bytes` for python2 compatibility. However, python3 forces two arguments to `bytes`, one of which is the encoding. Suggestions on how to make this better?

Reply:
it should be fine as

```python
if isinstance(base, six.text_type):
    base = base.encode(_utils._get_encoding())
```